File: app/src/main/java/com/huck/biblequiz/data/repository/BibleRepository.kt
package com.huck.biblequiz.data.repository

import android.content.Context
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.huck.biblequiz.data.local.dao.BookDao
import com.huck.biblequiz.data.local.dao.VerseDao
import com.huck.biblequiz.data.local.entity.BookEntity
import com.huck.biblequiz.data.local.entity.VerseEntity
import com.huck.biblequiz.data.remote.BollsApiService
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.zip.GZIPInputStream
import javax.inject.Inject
import javax.inject.Singleton

// ── Bundled-asset data model ──────────────────────────────────────────────────

private data class BundledBook(
    val bookId: Int,
    val name: String,
    val chapters: Int
)

private data class BundledVerse(
    val pk: Int,
    val bookId: Int,
    val chapter: Int,
    val verse: Int,
    val text: String
)

private data class BundledBible(
    val books: List<BundledBook>,
    val verses: List<BundledVerse>
)

// ── Repository ────────────────────────────────────────────────────────────────

@Singleton
class BibleRepository @Inject constructor(
    @ApplicationContext private val context: Context,
    private val api: BollsApiService,
    private val bookDao: BookDao,
    private val verseDao: VerseDao
) {

    suspend fun getBooks(): List<BookEntity> {
        val cached = bookDao.getAllBooks()
        if (cached.isNotEmpty()) return cached

        // Load from bundled asset (no network required)
        loadBundledBible()
        val afterLoad = bookDao.getAllBooks()
        if (afterLoad.isNotEmpty()) return afterLoad

        // Fallback to network if asset somehow fails
        val remote = api.getBooks()
        val entities = remote.map { BookEntity(it.bookId, it.name, it.chapters) }
        bookDao.insertBooks(entities)
        return entities
    }

    suspend fun getChapterVerses(bookId: Int, chapter: Int): List<VerseEntity> {
        val cached = verseDao.getVerses(bookId, chapter)
        if (cached.isNotEmpty()) return cached

        // Data should already be loaded from bundle; fallback to network
        val remote = api.getChapter(bookId, chapter)
        val entities = remote.map {
            VerseEntity(pk = it.pk, bookId = bookId, chapter = chapter,
                        verse = it.verse, text = it.text)
        }
        verseDao.insertVerses(entities)
        return entities
    }

    private suspend fun loadBundledBible() = withContext(Dispatchers.IO) {
        try {
            val json = context.assets.open("bible_nkjv.json.gz").use { raw ->
                BufferedReader(InputStreamReader(GZIPInputStream(raw))).readText()
            }
            val bible = Gson().fromJson(json, BundledBible::class.java)

            val bookEntities = bible.books.map { BookEntity(it.bookId, it.name, it.chapters) }
            bookDao.insertBooks(bookEntities)

            // Insert verses in per-book batches to keep memory use bounded
            val byBook = bible.verses.groupBy { it.bookId }
            for ((_, verses) in byBook) {
                val entities = verses.map {
                    VerseEntity(pk = it.pk, bookId = it.bookId, chapter = it.chapter,
                                verse = it.verse, text = it.text)
                }
                verseDao.insertVerses(entities)
            }
        } catch (_: Exception) {
            // Asset load failed — network fallback will handle it
        }
    }
}

File: ios/BeSpecific/Data/Repository/BibleRepository.swift
import Foundation
import SwiftData

// ── Bundled-asset data model ──────────────────────────────────────────────────

private struct BundledBook: Codable {
    let bookId: Int
    let name: String
    let chapters: Int
}

private struct BundledVerse: Codable {
    let pk: Int
    let bookId: Int
    let chapter: Int
    let verse: Int
    let text: String
}

private struct BundledBible: Codable {
    let books: [BundledBook]
    let verses: [BundledVerse]
}

// ── Repository ────────────────────────────────────────────────────────────────

@MainActor
final class BibleRepository {
    private let api: BollsAPIService
    private let modelContext: ModelContext

    init(api: BollsAPIService, modelContext: ModelContext) {
        self.api = api
        self.modelContext = modelContext
    }

    func getBooks() async throws -> [BookEntity] {
        let cached = try modelContext.fetch(FetchDescriptor<BookEntity>(
            sortBy: [SortDescriptor(\.bookId)]
        ))
        if !cached.isEmpty { return cached }

        // Load from bundled JSON — no network required
        if let loaded = try? loadBundledBible() {
            return loaded
        }

        // Fallback to network if bundle somehow missing
        let remote = try await api.getBooks()
        let entities = remote.map { BookEntity(bookId: $0.bookid, name: $0.name, chapters: $0.chapters) }
        for entity in entities { modelContext.insert(entity) }
        try modelContext.save()
        return entities
    }

    func getChapterVerses(bookId: Int, chapter: Int) async throws -> [VerseEntity] {
        let predicate = #Predicate<VerseEntity> { $0.bookId == bookId && $0.chapter == chapter }
        let descriptor = FetchDescriptor<VerseEntity>(
            predicate: predicate, sortBy: [SortDescriptor(\.verse)]
        )
        let cached = try modelContext.fetch(descriptor)
        if !cached.isEmpty { return cached }

        // Should already be in store from bundle load; fallback to network
        let remote = try await api.getChapter(bookId: bookId, chapter: chapter)
        let entities = remote.map {
            VerseEntity(pk: $0.pk, bookId: bookId, chapter: chapter,
                        verse: $0.verse, text: TextCleaner.clean($0.text))
        }
        for entity in entities { modelContext.insert(entity) }
        try modelContext.save()
        return entities
    }

    // ── Bundle load ───────────────────────────────────────────────────────────

    private func loadBundledBible() throws -> [BookEntity] {
        guard let url = Bundle.main.url(forResource: "bible_nkjv", withExtension: "json") else {
            throw NSError(domain: "BibleRepository", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "bible_nkjv.json not found in bundle"])
        }

        let data = try Data(contentsOf: url)
        let bible = try JSONDecoder().decode(BundledBible.self, from: data)

        let bookEntities = bible.books.map {
            BookEntity(bookId: $0.bookId, name: $0.name, chapters: $0.chapters)
        }
        for entity in bookEntities { modelContext.insert(entity) }

        // Insert verses grouped by book to keep memory pressure manageable
        let byBook = Dictionary(grouping: bible.verses, by: \.bookId)
        for (_, verses) in byBook {
            for v in verses {
                modelContext.insert(VerseEntity(
                    pk: v.pk, bookId: v.bookId, chapter: v.chapter,
                    verse: v.verse, text: v.text
                ))
            }
        }

        try modelContext.save()
        return bookEntities
    }
}
